keps/sig-architecture/20191001-prevent-permabeta.md

```diff
@@ -0,0 +1,140 @@
+
+Ensure the TOC is wrapped with <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code> tags, and then generate with `hack/update-toc.sh`.
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+APIs and features should not languish in beta.  They should take feedback and progress towards GA by either
+1. meeting GA criteria and getting promoted, or
+2. having a new beta and deprecating the previous beta
+  
+This must happen within six months (two releases).  If it does not,
```

Comment:
What about things like component config which depend on other APIs or other component configs?  
There is a transitive problem here where the top level dependency may be in a holding pattern due to it's leaf dependencies.

Reply:
I think 9-12 is more reasonable than 6 months. A reasonable policy would allow for features to apply for an extended beta. Perhaps with that added 6 months is OK. If you're not ready to graduate in 6 months you need to provide a justification not to be subject to the "beta expiration" policy.

Reply:
I echo @timothysc comment.  I would like to understand how this maps to feature gates for component config that impact the behavior of a specific component but not necessarily the end-user facing API.  For example, a feature like topology manager that enables a behavior in a kubelet, but does not have an exposure on the end-user facing API should probably be treated differently.  Other integration points between Kubernetes and components (CRI and CSI) probably should be called out separately and would follow a different path.

Reply:
+1 to the above. Getting user feedback is important for graduation to GA, and I consider 6mo the lower bound for getting feedback. This is especially true for features geared towards large enterprise customers, as those tend to lag several releases.

I think it's more important to see _progress_ every 6 months. In other words, as long as the feature is in beta, it should see regular improvements.